Fix ranges for range vars

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105055
* #102611
* __->__ #104987
* #104975

Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

lezcano committed:
Fix ranges for range vars

Ranges are inclusive on both ends...

We take this chance to delete a stale comment

[ghstack-poisoned]

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano commented on torch/_inductor/bounds.py:
```diff
@@ -19,7 +19,7 @@ class BoundVars:
         self.loop_body = loop_body
         self.replacement_vals = {
-            k: ValueRanges(0, v if not free_symbols(v) else math.inf)
+            k: ValueRanges(0, v - 1 if not free_symbols(v) else math.inf)
```
@ezyang should this be
```suggestion
            k: ValueRanges(0, v - 1) if not free_symbols(v) else ValueRanges(2, math.inf)
```
?

eellison replied:
The >= 2 assumption was used to reduce amount of (empirically) unnecessary guards. I don't think the same logic applies here, might affect correctness of generating code.

lezcano replied:
Can we have a dynamic size of size 1? I thought it would be specialised.

ezyang replied:
I don't really know what `replacement_vals` here is used for, so I am deferring to @eellison

eellison replied:
I think you can compile on dynamic, then pass in 1 later. That use case is why we have the assumption of >2 in tracing, to avoid recompilation, afaik.

lezcano replied:
If you pass a 0 or a 1 you recompile. Printing `loop_body.var_ranges` here and executing
```python
@torch.compile(dynamic=True)
def fn(a, b):
    return a[b]

x = torch.randn(8)
b = torch.tensor([1,1,2])
fn(x, b)
b = torch.zeros(2).long()
fn(x, b)
b = torch.zeros(1).long()
fn(x, b)
```
prints:
```
{z0: s1}
{}
```
i.e. first compilation, no recompilation with tensor of size 2, then compile again with tensor of size 1 (specified).

eellison replied:
You're right. I printed out the guards and I see the >=2 guards. 

I think this is true to the extent that we are not doing aot_inductor export. @desertfire do you know what the plan is for exporting wrt/guards on inputs, do we just ignore them ? do we have any plans to do multiple compilations for different guards ?

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

eellison reviewed (approved):
looks good although I think we have to be careful about the >= 2 assumption if we ever compile something like `nonzero`, and i also want to hear about export situation.

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

lezcano committed:
Update on "Fix ranges for range vars"


Ranges are inclusive on both ends...

We take this chance to delete a stale comment

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

eellison commented on torch/_inductor/bounds.py:
```diff
@@ -19,7 +19,9 @@ class BoundVars:
-            k: ValueRanges(0, v if not free_symbols(v) else math.inf)
+            k: ValueRanges(0, v - 1)
+            if not free_symbols(v)
+            else ValueRanges(2, math.inf)
```
https://github.com/pytorch/pytorch/blob/085dbfe2a984e4e49852bb34887d33b14c46c76d/torch/_inductor/utils.py#L392-L398

Do we not have the problem that some of these are shape symbols and some are index variables ?

lezcano replied:
@ezyang first put the `[0,inf]`patch, and he said that's alright because this is coming from `var_ranges` ¯\_(ツ)_/¯

ezyang replied:
Zero-inf is always safe, both shape symbols and index variables are non-negative. Two-inf is way more aggro :P

lezcano replied:
I assume that `var_ranges` will just contain shape symbols, right?

ezyang replied:
That I don't know. You'll have to read the code and find out, or ask @eellison :P

lezcano replied:
@ezyang You wrote that code in https://github.com/pytorch/pytorch/pull/103470

lezcano replied:
Probably a fix à la https://github.com/pytorch/pytorch/pull/105155 would be better here to be on the safe side. I'll put a PR up later today.

ezyang replied:
I'm not sure what you're referring to. In https://github.com/pytorch/pytorch/pull/103470 I did not touch bounds.py nor utils.py? I did add this test https://github.com/pytorch/pytorch/pull/103470/files#diff-e869ca39e024fc20e59f09f627ea4e3ce43aa664ebe9bc40d6b8ca2f201e5ff6R108 but the motivation was just to make sure we didn't error indexing into vars_ranges with a symbol that wasn't in the map.

lezcano replied:
`bounds.py` did not exist back then, it was part of `optimize_indexing.py`..

This code up here corresponds to what in that PR was
```python
        for k, v in indices_ranges.items():
            if free_symbols(v):
                v = math.inf
            self.replace_indirect(k, ValueRanges(0, v))
```
Note that the implementation of `replace_index` is just setting `replacement_vals`
```python
    def replace_indirect(self, old, new):
        """Swap in a variable used in indirect indexing"""
        assert isinstance(new, ValueRanges)
        self.replacement_vals[old] = new
```
which is exactly the dictionary we're setting here.

ezyang replied:
None of us are objecting to the inf upper bound. The entire question is around whether or not a 2 lower bound is permissible. I don't know how to reason out that it is OK.
